proposals/p2173.md

```diff
@@ -294,94 +707,76 @@ interface Example {
 ```
 
-Rewrite constraints thereby give us a deterministic, terminating type
-canonicalization mechanism for associated constants: in `A.B`, if the type of
-`A` specifies that `.B == C`, then `A.B` is replaced by `C`. Equality of types
-constrained in this way is transitive.
-
-Note that given `T:! C where .R = i32` can result in a type `T.R` whose behavior
-is different from the behavior of `T.R` given `T:! C`. For example, member
-lookup into `T.R` can find different results and operations can therefore have
-different behavior. However, this does not violate coherence because `C` and
-`C where .R = i32` don't differ by merely having more type information; rather,
-they are different types that have an isomorphic set of values, somewhat like
-`i32` and `u32`. An `=` constraint is not merely learning a new fact about a
-type, it is requesting different behavior.
-
-### Combining constraints with `&`
+#### Termination
 
-Suppose we have `X = C where .R = A` and `Y = C where .R = B`. What should
-`C & X` produce? What should `X & Y` produce?
+Each of the above steps performs at most one rewrite, and doesn't introduce any
+new recursive type-checking steps, so should not introduce any new forms of
+non-termination. Rewrite constraints thereby give us a deterministic,
+terminating type canonicalization mechanism for associated constants: in `A.B`,
+if the type of `A` specifies that `.B == C`, then `A.B` is replaced by `C`.
+Equality of types constrained in this way is transitive.
 
-We could perhaps say that `X & Y` results in a constraint where the type of `R`
-has the union of the interface of `A` and the interface of `B`, and that `C & X`
-similarly results in a constraint where the type of `R` has the union of the
-interface of `A` and the interface originally specified by `C`. However, this
-proposal suggests a simpler rule:
-
--   Combining two rewrite rules with different rewrite targets results in a
-    constraint where the associated constant is ambiguous. Given `T:! X & Y`,
-    the type expression `T.R` is ambiguous between a rewrite to `A` and a
-    rewrite to `B`. But given `T:! X & X`, `T.R` is unambiguously rewritten to
-    `A`.
--   Combining a constraint with a rewrite rule with a constraint with no rewrite
-    rule preserves the rewrite rule. For example, supposing that
-    `interface Container` extends `interface Iterable`, and `Iterable` has an
-    associated constant `Element`, the constraint
-    `Container & (Iterable where .Element = i32)` is the same as the constraint
-    `(Container & Iterable) where .Element = i32` which is the same as the
-    constraint `Container where .Element = i32`.
+However, some existing forms of non-termination may remain, such as substitution
+into a named constraint recursively triggering another such substitution, or
+template instantiation triggering another template instantiation. Such cases
+will need to be detected and handled in some way, such as by a depth limit, but
```

Comment:
Template instantiation triggering more template instantiation is a real case where I would expect this can happen, just as it does in C++. I think we might have handled the other cases I was thinking about here with your recent constraint completeness proposal, so I've removed mention of that. We also need to finalize the rules for preventing infinite regress when `impl` selection recursively performs more `impl` selection.